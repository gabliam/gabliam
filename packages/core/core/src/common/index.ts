export * from './utils';
export * from './interfaces';
