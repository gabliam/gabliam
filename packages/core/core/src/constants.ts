/** Type in registry */
export const TYPE = {
  Service: 'ServiceType',
  Config: 'ConfigType',
  PreDestroy: 'PreDestroy',
};

export const METADATA_KEY = {
  bean: '_bean',
  onMissingBean: '_onMissingBean',
  init: '_init',
  beforeCreate: '_beforeCreate',
  value: '_value',
  register: '_register',
  scan: '_scan',
  config: '_config',
  service: '_service',
  plugin: '_plugin',
  injectContainer: '_inject_container',
  preDestroy: '_preDestroy',
};

/**
 * Inject container key in descriptor
 */
export const INJECT_CONTAINER_KEY = '$$container';

export const ORDER_CONFIG = {
  Core: 50,
  Plugin: 100,
  Config: 150,
};

export const ERRORS_MSGS = {
  DUPLICATED_DECORATOR: `Cannot apply decorator multiple times.`,
  DUPLICATED_CONFIG_DECORATOR: `Cannot apply @Config, @PluginConfig or @CorePlugin decorator multiple times.`,
  DUPLICATED_PLUGIN_DECORATOR: `Cannot apply @Plugin decorator multiple times.`,
  DUPLICATED_APPLICATION_DECORATOR: `Cannot apply @Application decorator multiple times.`,
};

export const APP_CONFIG = Symbol('APP_CONFIG');

export const CORE_CONFIG = Symbol('CORE_CONFIG');

export const VALUE_EXTRACTOR = Symbol('ValueExtractor');
