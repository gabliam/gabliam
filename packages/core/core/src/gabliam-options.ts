import { GabliamConfig } from './interfaces';
import * as _ from 'lodash';

const DEFAULT_CONFIG: GabliamConfig = {
  config: process.env.GABLIAM_CONFIG_PATH,
};

export const getGabliamConfig = (
  options?: Partial<GabliamConfig> | string
): GabliamConfig => {
  if (options === undefined) {
    return DEFAULT_CONFIG;
  } else {
    if (_.isString(options)) {
      return {
        ...DEFAULT_CONFIG,
        config: options,
      };
    } else {
      return {
        ...DEFAULT_CONFIG,
        ...options,
      };
    }
  }
};
