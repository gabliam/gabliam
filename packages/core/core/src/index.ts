import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
