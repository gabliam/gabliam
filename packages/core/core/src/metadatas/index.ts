export * from './bean';
export { Config, PluginConfig } from './config';
export * from './init';
export * from './inject-container';
export * from './lifecycle';
export * from './plugin';
export * from './register';
export * from './scan';
export * from './service';
export * from './value';
