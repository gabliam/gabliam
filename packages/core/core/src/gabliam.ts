/* eslint-disable no-await-in-loop */
import { ExpressionParser } from '@gabliam/expression';
import d from 'debug';
import _ from 'lodash';
import { callInstance, toPromise } from './common';
import { APP_CONFIG, CORE_CONFIG, TYPE, VALUE_EXTRACTOR } from './constants';
import { Container, createContainer } from './container';
import {
  ConfigRegistry,
  GabliamAddPlugin,
  GabliamConfig,
  PreDestroyRegistry,
  ValueRegistry,
} from './interfaces';
import { LoaderConfig, LoaderModule } from './loaders';
import {
  Bean,
  BeforeCreate,
  Init,
  OnMissingBean,
  PreDestroy,
} from './metadatas';
import { PluginList } from './plugin-list';
import { reflection } from './reflection';
import { Registry } from './registry';
import { configureValueExtractor } from './value-extractor';
import { getGabliamConfig } from './gabliam-options';

const debug = d('Gabliam:core');

/**
 * Gabliam
 */
export class Gabliam {
  public container: Container = createContainer();

  public config: object;

  /**
   * Registry
   */
  public registry: Registry = new Registry();

  /**
   * Module loader
   */
  public loaderModule: LoaderModule = new LoaderModule();

  /**
   * Config loader
   */
  public loaderConfig: LoaderConfig = new LoaderConfig();

  /**
   * Plugin list
   */
  public pluginList: PluginList = new PluginList();

  public options: GabliamConfig;

  /**
   * Constructor
   * @param  {GabliamConfig|string} options?
   */
  constructor(options?: Partial<GabliamConfig> | string) {
    this.options = getGabliamConfig(options);
    /**
     * @TODO move in building phase
     */
    this.container.bind(CORE_CONFIG).toConstantValue(this.options);
  }

  /**
   * Add a plugin
   * @param  {GabliamPluginConstructor} plugin
   * @returns Gabliam
   */
  public addPlugin(plugin: GabliamAddPlugin): Gabliam {
    this.pluginList.add(plugin);
    return this;
  }

  /**
   * Add any plugins
   * @param  {GabliamPluginConstructor[]} ...plugins
   * @returns Gabliam
   */
  public addPlugins(...plugins: Array<GabliamAddPlugin>): Gabliam {
    for (const plugin of plugins) {
      this.addPlugin(plugin);
    }
    return this;
  }

  /**
   * Build gabliam
   * @returns Promise
   */
  public async build(): Promise<Gabliam> {
    this.pluginList.sort();

    /**
     * Load config file
     */
    await this._initializeConfig();

    /**
     * Loading phase
     */
    this.registry.addRegistry(
      this.loaderModule.load(this.options.scanPath, this.pluginList.plugins),
    );

    /**
     * Binding phase
     */
    await this._bind();

    /**
     * Config phase
     */
    await this._loadConfig();

    for (const plugin of this.pluginList.pluginsWithBuild) {
      await toPromise(plugin.build(this.container, this.registry));
    }

    return this;
  }

  /**
   * Build and start gabliam application
   * @returns Promise
   */
  async buildAndStart(): Promise<Gabliam> {
    await this.build();
    await this.start();
    return this;
  }

  /**
   * Stop and destroy gabliam application
   * @returns Promise
   */
  async stopAndDestroy(): Promise<Gabliam> {
    await this.stop();
    await this.destroy();
    return this;
  }

  /**
   * Start gabliam application
   *
   * call all plugin.start
   * @returns Promise
   */
  async start(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithStart) {
      await toPromise(plugin.start(this.container, this.registry));
    }

    return this;
  }

  /**
   * Stop gabliam application
   * call all plugin.stop
   *
   * @returns Promise
   */
  async stop(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithStop) {
      await toPromise(plugin.stop(this.container, this.registry));
    }

    return this;
  }

  /**
   * Destroy gabliam application
   * call all plugin.destroy
   *
   * @returns Promise
   */
  async destroy(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithDestroy) {
      await toPromise(plugin.destroy(this.container, this.registry));
    }

    const values = this.registry.get<PreDestroyRegistry>(TYPE.PreDestroy);

    const instanceToDestroy = async (
      value: ValueRegistry<PreDestroyRegistry>,
    ) => {
      if (value.options) {
        const instance = this.container.get(value.id);
        for (const preDestroy of value.options.preDestroys) {
          await callInstance(instance, preDestroy);
        }
      }
    };

    const p = [];
    for (const value of values) {
      p.push(instanceToDestroy(value));
    }

    await Promise.all(p);

    return this;
  }

  /**
   * Load config file and bind result in APP_CONFIG
   */
  private async _initializeConfig() {
    this.config = await this.loaderConfig.load(this.options.config);
    const config = this.config;
    this.container.bind(APP_CONFIG).toConstantValue(config);
    this.container
      .bind(VALUE_EXTRACTOR)
      .toConstantValue(configureValueExtractor(this.container));
    this.container
      .bind(ExpressionParser)
      .toConstantValue(new ExpressionParser(config));
  }

  /**
   * Binding phase
   * Binding of config and service classes.
   * call all plugin.bind
   */
  private async _bind() {
    debug('_bind');

    this.registry
      .getAllAutoBind()
      .forEach(({ id, target }) =>
        this.container.bind(id).to(target).inSingletonScope(),
      );

    for (const plugin of this.pluginList.pluginsWithBind) {
      await toPromise(plugin.bind(this.container, this.registry));
    }
  }

  /**
   * Config phase
   */
  private async _loadConfig() {
    debug('_loadConfig');

    // Get all config classes in registry
    let configsRegistry = this.registry.get<ConfigRegistry>(TYPE.Config);

    debug('configsRegistry', configsRegistry);
    if (configsRegistry) {
      configsRegistry = _.sortBy(configsRegistry, 'options.order');

      for (const { id: configId } of configsRegistry) {
        // Get config instance
        const confInstance = this.container.get<object>(configId);
        const ctor = confInstance.constructor;

        // get all bean metadata in config classes
        const beanMetadatas = reflection.propMetadataOfDecorator<Bean>(
          ctor,
          Bean,
        );

        // get on missing bean metadata
        const onMissingBeanMetadatas = reflection.propMetadataOfDecorator<OnMissingBean>(
          ctor,
          OnMissingBean,
        );

        const beforeCreateMetas = Object.keys(
          reflection.propMetadataOfDecorator(ctor, BeforeCreate),
        );

        const initMetadas = Object.keys(
          reflection.propMetadataOfDecorator(ctor, Init),
        );

        // call all beforeCreate method if exist
        if (Array.isArray(beforeCreateMetas)) {
          // No promise.all and await because order of beans are important
          for (const metada of beforeCreateMetas) {
            await callInstance(confInstance, metada);
          }
        }

        // If config has bean metadata
        if (Object.keys(beanMetadatas).length) {
          // No promise.all and await because order of beans are important
          for (const [key, beans] of Object.entries(beanMetadatas)) {
            const onMissingBeans = onMissingBeanMetadatas[key] || [];

            // by default all bean are missing
            let allMissing = true;

            // if there are onMissingBeans config, check if bean exist or no
            for (const onMissingBean of onMissingBeans) {
              try {
                this.container.getAll(onMissingBean.id);
                allMissing = false;
                // eslint-disable-next-line no-empty
              } catch {}
            }

            // if all beans are missing, so we create all beans
            if (allMissing) {
              for (const { id } of beans) {
                const bean = await callInstance(confInstance, key);
                this.container.bind(id).toConstantValue(bean);
                if (bean === undefined || bean.constructor === undefined) {
                  // eslint-disable-next-line no-continue
                  continue;
                }

                const preDestroys = Object.keys(
                  reflection.propMetadataOfDecorator(
                    bean.constructor,
                    PreDestroy,
                  ),
                );

                // bean can return undefined or can be a constant value
                if (preDestroys.length) {
                  this.registry.add(TYPE.PreDestroy, <ValueRegistry>{
                    id,
                    target: bean,
                    options: {
                      preDestroys,
                    },
                  });
                }
              }
            }
          }
        }

        if (Array.isArray(initMetadas)) {
          // No promise.all and await because order of beans are important
          for (const metada of initMetadas) {
            await callInstance(confInstance, metada);
          }
        }

        for (const plugin of this.pluginList.pluginWithConfig) {
          await toPromise(
            plugin.config(this.container, this.registry, confInstance),
          );
        }
      }
    }
    debug('_loadConfig end');
  }
}
