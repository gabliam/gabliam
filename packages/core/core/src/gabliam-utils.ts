import { TYPE } from './constants';
import { LoaderModule } from './loaders';
import { Registry } from './registry';
import { ValueRegistry } from './interfaces';
import { reflection } from './reflection';
import { Application } from './metadatas';
import { Gabliam } from './gabliam';
import { Type, toPromise } from './common';
import * as bluebird from 'bluebird';

/**
 * Class for build a gabliam application
 */
export const gabliamFinder = async (scanPath: string) => {
  const loaderModule = new LoaderModule();
  const registry = new Registry();
  registry.addRegistry(loaderModule.load(scanPath, []));
  return registry.get(TYPE.Application);
};

export const gabliamFindApp = async (scanPath: string, appName?: string) => {
  const list = await gabliamFinder(scanPath);
  let app: Type<any> | undefined;
  if (appName) {
    for (const value of list) {
      const metadata = reflection.annotationsOfDecorator<Application>(
        value.target,
        Application
      )[0];

      if (metadata) {
        const name = metadata.name ?? value.target.name;
        if (appName === name) {
          app = value.target;
          break;
        }
      }
    }
  } else {
    if (list.length > 1) {
      throw new Error(
        `Too many app (${list.length}) found. You must select an application.`
      );
    }

    if (list[0]) {
      app = list[0].target;
    }
  }

  if (app === undefined) {
    const errorMsg = appName
      ? `${appName} is not a gabliam application`
      : 'no application found';
    throw new Error(errorMsg);
  }

  return app;
};

const isValueRegistry = (val: any): val is ValueRegistry<any> => {
  return (
    val &&
    typeof val === 'object' &&
    val.hasOwnProperty('id') &&
    val.hasOwnProperty('target') &&
    val.hasOwnProperty('autoBind')
  );
};

export const gabliamBuilder = async <T = any>(
  clazzOrValue: ValueRegistry<T> | Type<any>
) => {
  let clazz: any = clazzOrValue;
  if (isValueRegistry(clazzOrValue)) {
    clazz = clazzOrValue.target;
  }

  const application = reflection.annotationsOfDecorator<Application>(
    clazz,
    Application
  )[0];
  if (application) {
    const plugins = await bluebird
      .filter(application.plugins, async item => toPromise(item.condition()))
      .map(item => {
        if (typeof item.plugin === 'string') {
          return require(item.plugin).default;
        }
        return item.plugin;
      });
    return new Gabliam(application.gabliamConfig).addPlugins(...plugins);
  }

  throw new Error(`${clazzOrValue} is not a gabliam application`);
};
