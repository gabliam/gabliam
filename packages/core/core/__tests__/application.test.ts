// tslint:disable:one-line
// tslint:disable:no-unused-expression
import { gabliamFinder, gabliamBuilder, gabliamFindApp } from '../src';
import { resolve } from 'path';

describe('gabliamFinder', () => {
  describe('one application', () => {
    test('app without config', async () => {
      const apps = await gabliamFinder(
        resolve(__dirname, './fixtures/applications/one/withoutconfig')
      );
      expect(apps).toMatchSnapshot();
      const gab = await gabliamBuilder(apps[0]);
      expect(gab).toMatchSnapshot();
    });

    test('app with config', async () => {
      const apps = await gabliamFinder(
        resolve(__dirname, './fixtures/applications/one/withconfig')
      );
      expect(apps).toMatchSnapshot();
      const gab = await gabliamBuilder(apps[0]);
      expect(gab).toMatchSnapshot();
    });

    test('no app found', async () => {
      const apps = await gabliamFinder(
        resolve(__dirname, './fixtures/applications/one/gabliam')
      );
      expect(apps).toMatchSnapshot();
      expect(gabliamBuilder(apps[0])).rejects.toMatchSnapshot();
    });
  });

  describe('2 applications', () => {
    test('app', async () => {
      const apps = await gabliamFinder(
        resolve(__dirname, './fixtures/applications/two')
      );
      expect(apps).toMatchSnapshot();
      expect(gabliamBuilder(apps[0])).resolves.toMatchSnapshot();
      expect(gabliamBuilder(apps[1])).resolves.toMatchSnapshot();
    });
  });
});

describe('gabliamFindApp', () => {
  describe('without appName', () => {
    describe('one application', () => {
      test('app without config', async () => {
        const apps = await gabliamFindApp(
          resolve(__dirname, './fixtures/applications/one/withoutconfig')
        );
        expect(apps).toMatchSnapshot();
        const gab = await gabliamBuilder(apps);
        expect(gab).toMatchSnapshot();
      });

      test('app with config', async () => {
        const apps = await gabliamFindApp(
          resolve(__dirname, './fixtures/applications/one/withconfig')
        );
        expect(apps).toMatchSnapshot();
        const gab = await gabliamBuilder(apps);
        expect(gab).toMatchSnapshot();
      });

      test('no app found', async () => {
        expect(
          gabliamFindApp(
            resolve(__dirname, './fixtures/applications/one/gabliam')
          )
        ).rejects.toMatchSnapshot();
      });
    });

    describe('2 applications', () => {
      test('too many', async () => {
        expect(
          gabliamFindApp(resolve(__dirname, './fixtures/applications/two'))
        ).rejects.toMatchSnapshot();
      });
    });
  });

  describe('with appName', () => {
    describe('one application', () => {
      test('app without config', async () => {
        const apps = await gabliamFindApp(
          resolve(__dirname, './fixtures/applications/one/withoutconfig'),
          'MyApp'
        );
        expect(apps).toMatchSnapshot();
        const gab = await gabliamBuilder(apps);
        expect(gab).toMatchSnapshot();
      });

      test('app with config', async () => {
        const apps = await gabliamFindApp(
          resolve(__dirname, './fixtures/applications/one/withconfig'),
          'MyApp'
        );
        expect(apps).toMatchSnapshot();
        const gab = await gabliamBuilder(apps);
        expect(gab).toMatchSnapshot();
      });

      test('no app found', async () => {
        expect(
          gabliamFindApp(
            resolve(__dirname, './fixtures/applications/one/withconfig'),
            'lolApp'
          )
        ).rejects.toMatchSnapshot();
      });
    });

    describe('2 applications', () => {
      test('app', async () => {
        const apps = await gabliamFindApp(
          resolve(__dirname, './fixtures/applications/two'),
          'MyApp'
        );
        expect(apps).toMatchSnapshot();
        expect(apps.name).toMatchSnapshot();
        expect(gabliamBuilder(apps)).resolves.toMatchSnapshot();
      });
    });
  });
});
