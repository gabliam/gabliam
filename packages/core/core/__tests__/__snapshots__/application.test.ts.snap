// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gabliamFindApp with appName 2 applications app 1`] = `[Function]`;

exports[`gabliamFindApp with appName 2 applications app 2`] = `"Application2"`;

exports[`gabliamFindApp with appName one application app with config 1`] = `[Function]`;

exports[`gabliamFindApp with appName one application app with config 2`] = `
Gabliam {
  "container": Container {
    "_bindingDictionary": Lookup {
      "_map": Map {
        Symbol(CORE_CONFIG) => Array [
          Binding {
            "activated": false,
            "cache": Object {
              "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig/config",
              "name": "MyApp",
              "plugins": Array [
                [Function],
              ],
              "scanPath": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig",
            },
            "constraint": [Function],
            "dynamicValue": null,
            "factory": null,
            "id": 15,
            "implementationType": null,
            "onActivation": [Function],
            "provider": null,
            "scope": "Transient",
            "serviceIdentifier": Symbol(CORE_CONFIG),
            "type": "ConstantValue",
          },
        ],
      },
    },
    "_metadataReader": MetadataReader {},
    "_middleware": [Function],
    "_snapshots": Array [],
    "bind": [Function],
    "id": 14,
    "options": Object {
      "autoBindInjectable": false,
      "defaultScope": "Transient",
      "skipBaseClassChecks": false,
    },
    "parent": null,
  },
  "loaderConfig": LoaderConfig {},
  "loaderModule": LoaderModule {
    "loadedFolder": Array [],
  },
  "options": Object {
    "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig/config",
    "name": "MyApp",
    "plugins": Array [
      [Function],
    ],
    "scanPath": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig",
  },
  "pluginList": PluginList {
    "_plugins": Array [
      Object {
        "dependencies": Array [],
        "name": "TypeOrmPlugin",
        "plugin": TypeOrmPlugin {},
      },
    ],
  },
  "registry": Registry {
    "registry": Map {},
  },
}
`;

exports[`gabliamFindApp with appName one application app without config 1`] = `[Function]`;

exports[`gabliamFindApp with appName one application app without config 2`] = `
Gabliam {
  "container": Container {
    "_bindingDictionary": Lookup {
      "_map": Map {
        Symbol(CORE_CONFIG) => Array [
          Binding {
            "activated": false,
            "cache": Object {
              "config": undefined,
              "name": "MyApp",
            },
            "constraint": [Function],
            "dynamicValue": null,
            "factory": null,
            "id": 13,
            "implementationType": null,
            "onActivation": [Function],
            "provider": null,
            "scope": "Transient",
            "serviceIdentifier": Symbol(CORE_CONFIG),
            "type": "ConstantValue",
          },
        ],
      },
    },
    "_metadataReader": MetadataReader {},
    "_middleware": [Function],
    "_snapshots": Array [],
    "bind": [Function],
    "id": 12,
    "options": Object {
      "autoBindInjectable": false,
      "defaultScope": "Transient",
      "skipBaseClassChecks": false,
    },
    "parent": null,
  },
  "loaderConfig": LoaderConfig {},
  "loaderModule": LoaderModule {
    "loadedFolder": Array [],
  },
  "options": Object {
    "config": undefined,
    "name": "MyApp",
  },
  "pluginList": PluginList {
    "_plugins": Array [],
  },
  "registry": Registry {
    "registry": Map {},
  },
}
`;

exports[`gabliamFindApp with appName one application no app found 1`] = `[Error: lolApp is not a gabliam application]`;

exports[`gabliamFindApp without appName 2 applications too many 1`] = `[Error: Too many app (2) found. You must select an application.]`;

exports[`gabliamFindApp without appName one application app with config 1`] = `[Function]`;

exports[`gabliamFindApp without appName one application app with config 2`] = `
Gabliam {
  "container": Container {
    "_bindingDictionary": Lookup {
      "_map": Map {
        Symbol(CORE_CONFIG) => Array [
          Binding {
            "activated": false,
            "cache": Object {
              "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig/config",
              "name": "MyApp",
              "plugins": Array [
                [Function],
              ],
              "scanPath": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig",
            },
            "constraint": [Function],
            "dynamicValue": null,
            "factory": null,
            "id": 11,
            "implementationType": null,
            "onActivation": [Function],
            "provider": null,
            "scope": "Transient",
            "serviceIdentifier": Symbol(CORE_CONFIG),
            "type": "ConstantValue",
          },
        ],
      },
    },
    "_metadataReader": MetadataReader {},
    "_middleware": [Function],
    "_snapshots": Array [],
    "bind": [Function],
    "id": 10,
    "options": Object {
      "autoBindInjectable": false,
      "defaultScope": "Transient",
      "skipBaseClassChecks": false,
    },
    "parent": null,
  },
  "loaderConfig": LoaderConfig {},
  "loaderModule": LoaderModule {
    "loadedFolder": Array [],
  },
  "options": Object {
    "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig/config",
    "name": "MyApp",
    "plugins": Array [
      [Function],
    ],
    "scanPath": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig",
  },
  "pluginList": PluginList {
    "_plugins": Array [
      Object {
        "dependencies": Array [],
        "name": "TypeOrmPlugin",
        "plugin": TypeOrmPlugin {},
      },
    ],
  },
  "registry": Registry {
    "registry": Map {},
  },
}
`;

exports[`gabliamFindApp without appName one application app without config 1`] = `[Function]`;

exports[`gabliamFindApp without appName one application app without config 2`] = `
Gabliam {
  "container": Container {
    "_bindingDictionary": Lookup {
      "_map": Map {
        Symbol(CORE_CONFIG) => Array [
          Binding {
            "activated": false,
            "cache": Object {
              "config": undefined,
            },
            "constraint": [Function],
            "dynamicValue": null,
            "factory": null,
            "id": 5,
            "implementationType": null,
            "onActivation": [Function],
            "provider": null,
            "scope": "Transient",
            "serviceIdentifier": Symbol(CORE_CONFIG),
            "type": "ConstantValue",
          },
        ],
      },
    },
    "_metadataReader": MetadataReader {},
    "_middleware": [Function],
    "_snapshots": Array [],
    "bind": [Function],
    "id": 4,
    "options": Object {
      "autoBindInjectable": false,
      "defaultScope": "Transient",
      "skipBaseClassChecks": false,
    },
    "parent": null,
  },
  "loaderConfig": LoaderConfig {},
  "loaderModule": LoaderModule {
    "loadedFolder": Array [],
  },
  "options": Object {
    "config": undefined,
  },
  "pluginList": PluginList {
    "_plugins": Array [],
  },
  "registry": Registry {
    "registry": Map {},
  },
}
`;

exports[`gabliamFindApp without appName one application app without config 3`] = `
Gabliam {
  "container": Container {
    "_bindingDictionary": Lookup {
      "_map": Map {
        Symbol(CORE_CONFIG) => Array [
          Binding {
            "activated": false,
            "cache": Object {
              "config": undefined,
              "name": "MyApp",
            },
            "constraint": [Function],
            "dynamicValue": null,
            "factory": null,
            "id": 7,
            "implementationType": null,
            "onActivation": [Function],
            "provider": null,
            "scope": "Transient",
            "serviceIdentifier": Symbol(CORE_CONFIG),
            "type": "ConstantValue",
          },
        ],
      },
    },
    "_metadataReader": MetadataReader {},
    "_middleware": [Function],
    "_snapshots": Array [],
    "bind": [Function],
    "id": 6,
    "options": Object {
      "autoBindInjectable": false,
      "defaultScope": "Transient",
      "skipBaseClassChecks": false,
    },
    "parent": null,
  },
  "loaderConfig": LoaderConfig {},
  "loaderModule": LoaderModule {
    "loadedFolder": Array [],
  },
  "options": Object {
    "config": undefined,
    "name": "MyApp",
  },
  "pluginList": PluginList {
    "_plugins": Array [],
  },
  "registry": Registry {
    "registry": Map {},
  },
}
`;

exports[`gabliamFindApp without appName one application app without config 4`] = `
Gabliam {
  "container": Container {
    "_bindingDictionary": Lookup {
      "_map": Map {
        Symbol(CORE_CONFIG) => Array [
          Binding {
            "activated": false,
            "cache": Object {
              "config": undefined,
              "name": "MyApp",
            },
            "constraint": [Function],
            "dynamicValue": null,
            "factory": null,
            "id": 9,
            "implementationType": null,
            "onActivation": [Function],
            "provider": null,
            "scope": "Transient",
            "serviceIdentifier": Symbol(CORE_CONFIG),
            "type": "ConstantValue",
          },
        ],
      },
    },
    "_metadataReader": MetadataReader {},
    "_middleware": [Function],
    "_snapshots": Array [],
    "bind": [Function],
    "id": 8,
    "options": Object {
      "autoBindInjectable": false,
      "defaultScope": "Transient",
      "skipBaseClassChecks": false,
    },
    "parent": null,
  },
  "loaderConfig": LoaderConfig {},
  "loaderModule": LoaderModule {
    "loadedFolder": Array [],
  },
  "options": Object {
    "config": undefined,
    "name": "MyApp",
  },
  "pluginList": PluginList {
    "_plugins": Array [],
  },
  "registry": Registry {
    "registry": Map {},
  },
}
`;

exports[`gabliamFindApp without appName one application no app found 1`] = `[Error: no application found]`;

exports[`gabliamFinder 2 applications app 1`] = `
Array [
  Object {
    "autoBind": false,
    "id": [Function],
    "options": undefined,
    "target": [Function],
  },
  Object {
    "autoBind": false,
    "id": [Function],
    "options": undefined,
    "target": [Function],
  },
]
`;

exports[`gabliamFinder one application app with config 1`] = `
Array [
  Object {
    "autoBind": false,
    "id": [Function],
    "options": undefined,
    "target": [Function],
  },
]
`;

exports[`gabliamFinder one application app with config 2`] = `
Gabliam {
  "container": Container {
    "_bindingDictionary": Lookup {
      "_map": Map {
        Symbol(CORE_CONFIG) => Array [
          Binding {
            "activated": false,
            "cache": Object {
              "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig/config",
              "name": "MyApp",
              "plugins": Array [
                [Function],
              ],
              "scanPath": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig",
            },
            "constraint": [Function],
            "dynamicValue": null,
            "factory": null,
            "id": 3,
            "implementationType": null,
            "onActivation": [Function],
            "provider": null,
            "scope": "Transient",
            "serviceIdentifier": Symbol(CORE_CONFIG),
            "type": "ConstantValue",
          },
        ],
      },
    },
    "_metadataReader": MetadataReader {},
    "_middleware": [Function],
    "_snapshots": Array [],
    "bind": [Function],
    "id": 2,
    "options": Object {
      "autoBindInjectable": false,
      "defaultScope": "Transient",
      "skipBaseClassChecks": false,
    },
    "parent": null,
  },
  "loaderConfig": LoaderConfig {},
  "loaderModule": LoaderModule {
    "loadedFolder": Array [],
  },
  "options": Object {
    "config": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig/config",
    "name": "MyApp",
    "plugins": Array [
      [Function],
    ],
    "scanPath": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/applications/one/withconfig",
  },
  "pluginList": PluginList {
    "_plugins": Array [
      Object {
        "dependencies": Array [],
        "name": "TypeOrmPlugin",
        "plugin": TypeOrmPlugin {},
      },
    ],
  },
  "registry": Registry {
    "registry": Map {},
  },
}
`;

exports[`gabliamFinder one application app without config 1`] = `
Array [
  Object {
    "autoBind": false,
    "id": [Function],
    "options": undefined,
    "target": [Function],
  },
]
`;

exports[`gabliamFinder one application app without config 2`] = `
Gabliam {
  "container": Container {
    "_bindingDictionary": Lookup {
      "_map": Map {
        Symbol(CORE_CONFIG) => Array [
          Binding {
            "activated": false,
            "cache": Object {
              "config": undefined,
              "name": "MyApp",
            },
            "constraint": [Function],
            "dynamicValue": null,
            "factory": null,
            "id": 1,
            "implementationType": null,
            "onActivation": [Function],
            "provider": null,
            "scope": "Transient",
            "serviceIdentifier": Symbol(CORE_CONFIG),
            "type": "ConstantValue",
          },
        ],
      },
    },
    "_metadataReader": MetadataReader {},
    "_middleware": [Function],
    "_snapshots": Array [],
    "bind": [Function],
    "id": 0,
    "options": Object {
      "autoBindInjectable": false,
      "defaultScope": "Transient",
      "skipBaseClassChecks": false,
    },
    "parent": null,
  },
  "loaderConfig": LoaderConfig {},
  "loaderModule": LoaderModule {
    "loadedFolder": Array [],
  },
  "options": Object {
    "config": undefined,
    "name": "MyApp",
  },
  "pluginList": PluginList {
    "_plugins": Array [],
  },
  "registry": Registry {
    "registry": Map {},
  },
}
`;

exports[`gabliamFinder one application no app found 1`] = `Array []`;

exports[`gabliamFinder one application no app found 2`] = `[Error: undefined is not a gabliam application]`;
